flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateSerializer.java

```diff
@@ -0,0 +1,199 @@
+}
+
+/**
+ * Wrapper around various buffers to receive channel state data.
+ */
+@Internal
+@NotThreadSafe
+interface ChannelStateByteBuffer {
+
+	boolean isWritable();
+
+	int writableBytes();
+
+	int writeBytes(InputStream input, int bytesToRead) throws IOException;
+
+	static ChannelStateByteBuffer wrap(Buffer buffer) {
+		return new ChannelStateByteBuffer() {
+
+			private ByteBuf byteBuf = buffer.asByteBuf();
+
+			@Override
+			public boolean isWritable() {
+				return byteBuf.isWritable();
+			}
+
+			@Override
+			public int writableBytes() {
+				return byteBuf.writableBytes();
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int bytesToRead) throws IOException {
+				return byteBuf.writeBytes(input, Math.min(bytesToRead, byteBuf.writableBytes()));
+			}
+		};
+	}
+
+	static ChannelStateByteBuffer wrap(BufferBuilder bufferBuilder) {
+		final byte[] buf = new byte[1024];
+		return new ChannelStateByteBuffer() {
+			@Override
+			public boolean isWritable() {
+				return !bufferBuilder.isFull();
+			}
+
+			@Override
+			public int writableBytes() {
+				return bufferBuilder.writableBytes();
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int upToBytes) throws IOException {
+				int left = upToBytes;
+				for (int toRead = getToRead(left); toRead > 0; toRead = getToRead(left)) {
+					int read = input.read(buf, 0, toRead);
+					int copied = bufferBuilder.append(java.nio.ByteBuffer.wrap(buf, 0, read));
+					Preconditions.checkState(copied == read);
+					left -= read;
+				}
+				bufferBuilder.commit();
+				return upToBytes - left;
+			}
+
+			private int getToRead(int bytesToRead) {
+				return min(bytesToRead, min(buf.length, writableBytes()));
+			}
+		};
+	}
+
+	static ChannelStateByteBuffer wrap(byte[] bytes) {
+		int written = 0;
+		return new ChannelStateByteBuffer() {
+			@Override
+			public boolean isWritable() {
+				return written < bytes.length;
+			}
+
+			@Override
+			public int writableBytes() {
+				return bytes.length - written;
+			}
+
+			@Override
+			public int writeBytes(InputStream input, int bytesToRead) throws IOException {
+				return input.read(bytes, written, writableBytes());
+			}
+		};
+	}
+}
+
+class ChannelStateSerializerImpl implements ChannelStateSerializer, ChannelStateDeserializer {
+	private static final int LEN_SIZE = Integer.BYTES;
+	private static final int SERIALIZATION_VERSION = 0;
+
+	@Override
+	public void writeHeader(DataOutputStream dataStream) throws IOException {
+		dataStream.writeInt(SERIALIZATION_VERSION);
+	}
+
+	@Override
+	public void writeData(DataOutputStream stream, Buffer... flinkBuffers) throws IOException {
+		stream.writeInt(getSize(flinkBuffers));
+		for (Buffer buffer : flinkBuffers) {
+			ByteBuf nettyByteBuf = buffer.asByteBuf();
+			nettyByteBuf.getBytes(nettyByteBuf.readerIndex(), stream, nettyByteBuf.readableBytes()); // todo: review me
```

Comment:
looks okay. anything particular to check?

Reply:
I wanted to make sure that it's not required to update reader index here.